iml-agent/src/action_plugins/stratagem/action_warning.rs

```diff
@@ -61,54 +52,57 @@ pub fn write_records(
     txt_path.push(mailbox.to_string());
     txt_path.set_extension("txt");
 
-    let s = mailbox_client::get(mailbox)
-        .filter_map(|x| {
-            x.trim()
-                .split(' ')
-                .filter(|x| x != &"")
-                .last()
-                .map(|x| fidlist::FidListItem::new(x.into()))
+    let f = iml_fs::file_write_bytes(txt_path.clone()).await?;
+
+    let llapi = search_rootpath(fsname_or_mntpath).await?;
+
+    let mntpt = llapi.mntpt();
+
+    mailbox_client::get(mailbox)
+        .try_filter_map(|x| {
+            future::ok(
+                x.trim()
+                    .split(' ')
+                    .filter(|x| x != &"")
+                    .last()
+                    .map(|x| fidlist::FidListItem::new(x.into())),
+            )
         })
         .chunks(1000)
-        .instrument(span!(Level::INFO, "Incoming fids"));
-
-    iml_fs::file_write_bytes(txt_path.clone())
-        .from_err()
-        .and_then(|f| search_rootpath(fsname_or_mntpath).map(|llapi| (llapi, f)))
-        .and_then(|(llapi, f)| {
-            let mntpt = llapi.mntpt();
-
-            let s2 = s
-                .and_then(move |xs| {
-                    let llapi2 = llapi.clone();
-
-                    join_all(xs.into_iter().map(move |x| fid2path(llapi2.clone(), x.fid)))
-                })
-                .inspect(|_| debug!("Resolved 1000 Fids"))
-                .map(move |xs| {
-                    xs.into_iter()
-                        .filter_map(std::convert::identity)
-                        .map(|x| format!("{}/{}", mntpt, x))
-                        .collect()
-                })
-                .map(|xs: Vec<String>| -> bytes::BytesMut { xs.join("\n").into() })
-                .map(|mut x: bytes::BytesMut| {
-                    if !x.is_empty() {
-                        x.extend_from_slice(b"\n");
-                    }
-
-                    x.freeze()
-                });
-
-            f.sink_from_err().send_all(s2).map(drop)
+        .map(|xs| -> Result<Vec<_>, _> { xs.into_iter().collect() })
+        .and_then(|xs| {
+            let llapi2 = llapi.clone();
+
+            async move {
+                let xs =
+                    join_all(xs.into_iter().map(move |x| fid2path(llapi2.clone(), x.fid))).await;
+
+                Ok(xs)
+            }
+        })
+        .inspect(|_| debug!("Resolved 1000 Fids"))
+        .map_ok(move |xs| {
+            xs.into_iter()
+                .filter_map(std::convert::identity)
+                .map(|x| format!("{}/{}", mntpt, x))
+                .collect()
         })
-        .map(move |_| txt_path)
-        .instrument(span!(Level::INFO, "Fid writer",))
+        .map_ok(|xs: Vec<String>| -> bytes::BytesMut { xs.join("\n").into() })
+        .map_ok(|mut x: bytes::BytesMut| {
+            if !x.is_empty() {
+                x.extend_from_slice(b"\n");
+            }
+            x.freeze()
+        })
+        .forward(f.sink_err_into())
```

Comment:
What does `forward` do here?

Reply:
This forwards all bytes from the mailbox into the file.

Reply:
I see.